cirq/protocols/apply_mixture_protocol.py

```diff
@@ -0,0 +1,309 @@
+                dtype as the target tensor.
+            auxiliary_buffer1: Pre-allocated workspace with the same shape
+                and dtype as the target tensor.
+            left_axes: Which axes to multiply the left action of the mixture
+                upon.
+            right_axes: Which axes to multiply the right action of the mixture
+                upon.
+        """
+        self.target_tensor = target_tensor
+        self.out_buffer = out_buffer
+        self.auxiliary_buffer0 = auxiliary_buffer0
+        self.auxiliary_buffer1 = auxiliary_buffer1
+        self.left_axes = tuple(left_axes)
+        self.right_axes = tuple(right_axes)
+
+
+class SupportsApplyMixture(Protocol):
+    """An object that can efficiently implement a mixture."""
+
+    def _apply_mixture_(self, args: ApplyMixtureArgs
+        ) -> Union[np.ndarray, None, NotImplementedType]:
+        """Efficiently applies a mixture.
+
+        This method is given both the target tensor and workspace of the same
+        shape and dtype. The method then either performs inline modifications of
+        the target tensor and returns it, or writes its output into the
+        a workspace tensor and returns that. This signature makes it possible to
+        write specialized simulation methods that run without performing large
+        allocations, significantly increasing simulation performance.
+
+        Args:
+            args: A `cirq.ApplyMixtureArgs` object with the `args.target_tensor`
+                to operate on, an `args.available_workspace` buffer to use as
+                temporary workspace, and the `args.axes` of the tensor to target
+                with the unitary operation. Note that this method is permitted
+                (and in fact expected) to mutate `args.target_tensor` and
+                `args.available_workspace`.
+
+        Returns:
+            If the receiving object is not able to apply its unitary effect,
+            None or NotImplemented should be returned.
+
+            If the receiving object is able to work inline, it should directly
+            mutate `args.target_tensor` and then return `args.target_tensor`.
+            The caller will understand this to mean that the result is in
+            `args.target_tensor`.
+
+            If the receiving object is unable to work inline, it can write its
+            output over `args.available_buffer` and then return
+            `args.available_buffer`. The caller will understand this to mean
+            that the result is in `args.available_buffer` (and so what was
+            `args.available_buffer` will become `args.target_tensor` in the next
+            call, and vice versa).
+
+            The receiving object is also permitted to allocate a new
+            numpy.ndarray and return that as its result.
+        """
+
+
+def apply_mixture(val: Any,
+                  args: ApplyMixtureArgs,
+                  default: TDefault = RaiseTypeErrorIfNotProvided
```

Comment:
We haven't been property consistent about this everywhere, but force keyword args on the default value.

Reply:
done